Design Paywalls V2 reactive state adapter (#6794)
RevenueCat/purchases-ios: docs/superpowers/plans/2026-05-14-paywalls-v2-reactive-state-adapter.md

# Paywalls V2 Reactive State Adapter Implementation Plan

> **For agentic workers:** REQUIRED SUB-SKILL: Use superpowers:subagent-driven-development (recommended) or superpowers:executing-plans to implement this plan task-by-task. Steps use checkbox (`- [ ]`) syntax for tracking.

**Goal:** Build the first backward-compatible Paywalls V2 reactive state adapter without changing the existing Paywalls V2 JSON payload shape.

**Architecture:** Add an internal RevenueCatUI state pipeline scoped by paywall instance, paywall ID, component ID, and field. Preserve existing override semantics while migrating package selection and a first text-field projection through request/gate/commit state changes.

> **Tech Stack:** Swift, SwiftUI, Combine, XCTest, Nimble, existing RevenueCatUI Paywalls V2 models.

Author: ```suggestion
**Tech Stack:** Swift, SwiftUI, Combine, XCTest, existing RevenueCatUI Paywalls V2 models.
```

Create `RevenueCatUI/Templates/V2/State/PaywallStateKey.swift`:

```swift
import Foundation

#if !os(tvOS)

@available(iOS 15.0, macOS 12.0, tvOS 15.0, watchOS 8.0, *)
@_spi(Internal) public struct PaywallStateKey: Hashable, Sendable {

    @_spi(Internal) public struct Field: Hashable, Sendable, RawRepresentable {
        @_spi(Internal) public let rawValue: String

        @_spi(Internal) public init(rawValue: String) {
            self.rawValue = rawValue
        }

        @_spi(Internal) public static let rootSelectedPackageID = Field(rawValue: "paywall.root_selected_package_id")
        @_spi(Internal) public static let componentVisible = Field(rawValue: "component.visible")
        @_spi(Internal) public static let componentText = Field(rawValue: "component.text")
        @_spi(Internal) public static let componentColor = Field(rawValue: "component.color")
        @_spi(Internal) public static let componentFontWeight = Field(rawValue: "component.font_weight")
        @_spi(Internal) public static let componentPadding = Field(rawValue: "component.padding")
        @_spi(Internal) public static let componentMargin = Field(rawValue: "component.margin")

> 

Author: I am interested in making this 100% dynamic. Instead of hardcoding a preset amount of defaults, I would much rather use the overrides JSON to construct the keys themselves. 

For example:

```JSON
"overrides": [
  {
    "conditions": [
      { "type": "selected" }
    ],
    "properties": {
      "color": {
        "light": {
          "type": "hex",
          "value": "#2EBEE5FF"
        }
      },
      "formats": {
        "heic": "filled-circle-check.heic",
        "png": "filled-circle-check.png",
        "svg": "filled-circle-check.svg",
        "webp": "filled-circle-check.webp"
      },
      "icon_name": "filled-circle-check"
    }
  }
],
```

This would create the keys of `component.color` and `component_iconURL` and we would just write the reducer to generate the url from the icon name and formats.

And with custom variables:

```JSON
"overrides": [
  {
    "conditions": [
      {
        "operator": "=",
        "type": "variable_condition",
        "value": false,
        "variable": "is_screen_width_large"
      }
    ],
    "properties": { "visible": false }
  },
  {
    "conditions": [
      {
        "operator": "=",
        "type": "variable_condition",
        "value": true,
        "variable": "is_screen_width_large"
      }
    ],
    "properties": { "visible": true }
  }
],
```

we would get `component.visible` for the key

The goal being that we can introduce new functionality without having to make SDK code changes. Dynamism. Composition over Inheritance

            component: .init(
                paywallID: "paywall_a",
                componentID: "component_a",
                type: "text",
                name: nil
            ),
            field: field
        )
    }

}

private struct TestStateChangeDetails: PaywallStateChangeDetails {
    let source: String
}

#endif
```

- [ ] **Step 2: Run test to verify it fails**

Run:

```bash
swift test --filter PaywallStateStoreTests
```

Expected: fail to compile because store and mutation types do not exist.

- [ ] **Step 3: Add mutation and proposal types**

Mirror the `ios-state-example` shape here. The store must not own a mutable global handler. The handler is passed into each request from the SwiftUI environment or the package selection coordinator. A handler that receives a proposal must resolve it exactly once by calling `accept()`, `reject()`, or `replace(with:)`; unresolved proposals assert in DEBUG on deinit.

Use generic detail payloads. Reducers should inspect `change.details as? SomeConcreteDetails` and act only on the detail type they understand. This avoids a single catch-all details struct that grows a field for every future event source.

> Use a phantom stage type on state changes. Proposals carry `PaywallStateChange<PaywallStateChangeProposed>` and committed event streams emit `PaywallStateChange<PaywallStateChangeCommitted>`, which prevents accidentally passing a proposed change where a committed change is expected.

Author: ```suggestion
Use a phantom stage type on state changes. Proposals carry `PaywallStateChange<ChangeProposed>` and committed event streams emit `PaywallStateChange<ChangeCommitted>`, which prevents accidentally passing a proposed change where a committed change is expected.
```

Create `RevenueCatUI/Templates/V2/State/PaywallStateMutation.swift`:

```swift
import Foundation

#if !os(tvOS)

@available(iOS 15.0, macOS 12.0, tvOS 15.0, watchOS 8.0, *)
@_spi(Internal) public struct PaywallStateMutation: Hashable, Sendable {
    @_spi(Internal) public let key: PaywallStateKey
    @_spi(Internal) public let value: PaywallStateValue?

    @_spi(Internal)
    public init(key: PaywallStateKey, value: PaywallStateValue?) {
        self.key = key
        self.value = value
    }
}

@available(iOS 15.0, macOS 12.0, tvOS 15.0, watchOS 8.0, *)

> @_spi(Internal) public struct PaywallState: Sendable {

Author: ```suggestion
@_spi(Internal) extension PaywallState {
```

@available(iOS 15.0, macOS 12.0, tvOS 15.0, watchOS 8.0, *)
@_spi(Internal) public protocol PaywallStateChangeStage: Sendable {}

@available(iOS 15.0, macOS 12.0, tvOS 15.0, watchOS 8.0, *)
@_spi(Internal) public struct PaywallStateChangeProposed: PaywallStateChangeStage {
    @_spi(Internal) public init() {}
}

@available(iOS 15.0, macOS 12.0, tvOS 15.0, watchOS 8.0, *)
@_spi(Internal) public struct PaywallStateChangeCommitted: PaywallStateChangeStage {
    @_spi(Internal) public init() {}
}

@available(iOS 15.0, macOS 12.0, tvOS 15.0, watchOS 8.0, *)

> @_spi(Internal) public struct PaywallStateChange<Stage: PaywallStateChangeStage>: Sendable {

Author: ```suggestion
@_spi(Internal) public protocol PaywallStateChangeStage: Sendable {}

@_spi(Internal) public enum PaywallState {

    @available(iOS 15.0, macOS 12.0, tvOS 15.0, watchOS 8.0, *)
    @_spi(Internal) public protocol Details: Hashable, Sendable {}

    @available(iOS 15.0, macOS 12.0, tvOS 15.0, watchOS 8.0, *)

    @available(iOS 15.0, macOS 12.0, tvOS 15.0, watchOS 8.0, *)
    @_spi(Internal) public enum ChangeProposed: PaywallStateChangeStage {}

    @available(iOS 15.0, macOS 12.0, tvOS 15.0, watchOS 8.0, *)
    @_spi(Internal) public enum ChangeCommitted: PaywallStateChangeStage {}
    @available(iOS 15.0, macOS 12.0, tvOS 15.0, watchOS 8.0, *)

    @_spi(Internal) public struct Change<Stage: PaywallStateChangeStage>: Sendable {
}
```

        newValue: PaywallStateValue?,
        details: (any PaywallStateChangeDetails)?
    ) {
        self.key = key
        self.oldValue = oldValue
        self.newValue = newValue
        self.details = details
    }
}

@available(iOS 15.0, macOS 12.0, tvOS 15.0, watchOS 8.0, *)
@_spi(Internal) public struct PaywallStateMutationHandler {
    private let action: @Sendable (PaywallStateMutationProposal) -> Void

    @_spi(Internal)
    public init(action: @escaping @Sendable (PaywallStateMutationProposal) -> Void) {
        self.action = action
    }

    func callAsFunction(_ proposal: PaywallStateMutationProposal) {
        self.action(proposal)
    }
}

@available(iOS 15.0, macOS 12.0, tvOS 15.0, watchOS 8.0, *)
@_spi(Internal) public final class PaywallStateMutationProposal: @unchecked Sendable {
    @_spi(Internal) public let change: PaywallStateChange<PaywallStateChangeProposed>

    private let lock = NSLock()
    private var resolve: ((Resolution) -> Void)?

    init(
        change: PaywallStateChange<PaywallStateChangeProposed>,
        resolve: @escaping (Resolution) -> Void
    ) {
        self.change = change
        self.resolve = resolve

>     }

Author: ```suggestion
    init(
        change: PaywallState.Change<ChangeProposed>,
        resolve: @escaping (Resolution) -> Void
    ) {
        self.change = change
        self.resolve = resolve
    }
```

Create `RevenueCatUI/Templates/V2/State/PaywallStateStore.swift`:

```swift
import Combine
import Foundation

#if !os(tvOS)

@available(iOS 15.0, macOS 12.0, tvOS 15.0, watchOS 8.0, *)
final class PaywallStateStore: ObservableObject {

    private let lock = NSLock()
    private var subjects: [PaywallStateKey: CurrentValueSubject<PaywallStateValue?, Never>]
    private let resolvedEventsSubject = PassthroughSubject<
        PaywallStateChange<PaywallStateChangeCommitted>,
        Never
    >()

    init(initialValues: [PaywallStateKey: PaywallStateValue?] = [:]) {
        self.subjects = initialValues.mapValues { CurrentValueSubject<PaywallStateValue?, Never>($0) }
    }

    var resolvedEvents: AnyPublisher<PaywallStateChange<PaywallStateChangeCommitted>, Never> {
        self.resolvedEventsSubject.eraseToAnyPublisher()
    }

    func value(for key: PaywallStateKey) -> PaywallStateValue? {
        self.lock.withLock {
            self.subjects[key]?.value
        }
    }

    func publisher(for key: PaywallStateKey) -> AnyPublisher<PaywallStateValue?, Never> {
        self.subject(for: key).eraseToAnyPublisher()
    }

    func request(
        _ mutation: PaywallStateMutation,

>         details: (any PaywallStateChangeDetails)?,

Author: ```suggestion
        details: (any PaywallState.Details)?,
```

        mutationHandler: PaywallStateMutationHandler? = nil
    ) {
        let proposedChange = PaywallStateChange<PaywallStateChangeProposed>(
            key: mutation.key,
            oldValue: self.value(for: mutation.key),
            newValue: mutation.value,
            details: details
        )

        guard let mutationHandler else {
            self.commit(mutation, details: details)
            return
        }

        let proposal = PaywallStateMutationProposal(change: proposedChange) { [weak self] resolution in
            switch resolution {
            case .accept:
                self?.commit(mutation, details: details)
            case .reject:
                break
            case .replace(let replacement):
                self?.commit(replacement, details: details)
            }
        }
        mutationHandler(proposal)
    }

>     private func commit(_ mutation: PaywallStateMutation, details: (any PaywallStateChangeDetails)?) {

Author: ```suggestion
    private func commit(_ mutation: PaywallStateMutation, details: (any PaywallState.Details)?) {
```

        let subject = self.subject(for: mutation.key)
        let oldValue = subject.value
        guard oldValue != mutation.value else { return }

>         let change = PaywallStateChange<PaywallStateChangeCommitted>(

Author: ```suggestion
        let change = PaywallState.Change<PaywallState.ChangeCommitted>(
```

            self.subjects[key]?.value
        }
    }

    func publisher(for key: PaywallStateKey) -> AnyPublisher<PaywallStateValue?, Never> {
        self.subject(for: key).eraseToAnyPublisher()
    }

    func request(
        _ mutation: PaywallStateMutation,
        details: (any PaywallStateChange.Details)?,
        mutationHandler: PaywallStateMutationHandler? = nil
    ) {
        let proposedChange = PaywallStateChange.Event<PaywallStateChange.Proposed>(
            key: mutation.key,
            oldValue: self.value(for: mutation.key),
            newValue: mutation.value,
            details: details
        )

        guard let mutationHandler else {
            self.commit(mutation, details: details)
            return
        }

        let proposal = PaywallStateMutationProposal(change: proposedChange) { [weak self] resolution in
            switch resolution {
            case .accept:
                self?.commit(mutation, details: details)
            case .reject:
                break
            case .replace(let replacement):
                self?.commit(replacement, details: details)
            }
        }
        mutationHandler(proposal)
    }

>     private func commit(_ mutation: PaywallStateMutation, details: (any PaywallStateChange.Details)?) {

Author: High: commit reads and sends CurrentValueSubject values outside the lock. CurrentValueSubject is not a synchronization primitive, so concurrent requests can publish stale oldValue values or race commits. The store should keep canonical values under Lock or NSLock, update them atomically, then publish outside the lock.

```swift
try container.encode(id, forKey: .id)
```

Update initializers with a default for test ergonomics only if existing tests create models directly:

```swift
id: String = UUID().uuidString
```

Prefer explicit IDs in new tests. Preserve existing tests by providing a default value in source initializers.

For every partial component that can appear in override `properties`, add explicit `CodingKeys` if it currently relies on synthesized coding keys. Keep those coding keys at least `internal` so reducer code can reference `CodingKeys.<property>.stringValue` instead of duplicating serialized property names.

- [ ] **Step 4: Preserve raw override property names**

In `ComponentOverrides.swift`, keep the existing typed partial decode for compatibility, but also store a raw JSON property bag. The purpose is not to render unknown fields immediately; it is to generate state keys and expose unresolved properties to reducer code without waiting for a new SDK release.

Add a small JSON value type or reuse `PaywallStateJSONValue` if it lives in a module visible to `RevenueCat`. Because `Sources/Paywalls` cannot depend on `RevenueCatUI`, prefer a source-level type such as:

```swift

> @_spi(Internal) public struct PaywallComponentPropertyValue: Codable, Hashable, Sendable {

Author: High: This sketches PaywallComponentPropertyValue as Codable with private Storage, but there is no custom decoder. Synthesis will not decode arbitrary JSON scalars, arrays, and objects from the properties container, so raw override preservation will not work. PaywallStateJSONValue has the same issue. Add explicit init(from:) and encode(to:) implementations using single-value, keyed dynamic-key, and unkeyed containers.

            paywallID: paywallID,
            offeringIdentifier: "default",
            paywallRevision: 1,
            workflowPageID: nil
        )
    }

    private static var packageIdentity: PaywallComponentIdentity {
        packageIdentity(paywallID: "paywall_a", componentID: "package_monthly")
    }

    private static func packageIdentity(paywallID: String, componentID: String) -> PaywallComponentIdentity {
        PaywallComponentIdentity(
            paywallID: paywallID,
            componentID: componentID,
            type: "package",
            name: nil
        )
    }

}

#endif
```

- [ ] **Step 2: Run tests to verify they fail**

Run:

```bash
swift test --filter PaywallPackageSelectionCoordinatorTests
```

Expected: fail because coordinator does not exist.

- [ ] **Step 3: Implement coordinator**

Create `PaywallPackageSelectionCoordinator.swift`:

```swift
@_spi(Internal) import RevenueCat
import Foundation

#if !os(tvOS)

@available(iOS 15.0, macOS 12.0, tvOS 15.0, watchOS 8.0, *)
struct PaywallPackageSelectionDetails: PaywallStateChange.Details {
    let sourceComponentID: String
    let sheetComponentID: String?
    let reason: String?
}

> Details must remain plain `Sendable` values without actor isolation. Apply `@MainActor` to the side-effect coordinator that mutates `PackageContext`, not to the detail payload.

Author: Medium: This explanatory sentence is inside the Swift code block. If a worker copies the snippet literally, PaywallPackageSelectionCoordinator.swift will not compile. Move the note outside the code block or convert it into a Swift comment.

            packageID: package.identifier,
            field: .sheetSelectedPackageID(componentID: sheetComponentID),
            sourceComponent: sourceComponent,
            sheetComponentID: sheetComponentID
        )
    }

    func restoreRootSelectionAfterSheetDismiss() {
        let sheetComponentID = self.activeSheetComponentID
        self.activeSheetComponentID = nil
        if let sheetComponentID {
            let sheetKey = PaywallStateKey.paywall(
                scope: self.scope,
                field: .sheetSelectedPackageID(componentID: sheetComponentID)
            )
            self.store.request(
                .init(key: sheetKey, value: nil),
                details: PaywallPackageSelectionDetails(
                    sourceComponentID: "sheet_dismiss",
                    sheetComponentID: sheetComponentID,
                    reason: "dismiss"
                ),
                mutationHandler: self.mutationHandler
            )
        }
        let rootKey = PaywallStateKey.paywall(scope: self.scope, field: .rootSelectedPackageID)
        let rootPackageID = self.store.value(for: rootKey)?.packageID
        self.packageContext.update(

>             package: rootPackageID.flatMap { self.packagesByID[$0] } ?? self.defaultPackage,

Author: Medium: This restores sheet dismissal to the coordinator captured defaultPackage. Existing RootView behavior reads the current workflow default at dismissal time, so this can restore stale state if workflowPackageContext changes after coordinator creation. The coordinator should receive or query the current workflow selected package when dismissal happens.

            variableContext: self.packageContext.variableContext
        )
    }

    private func request(
        packageID: String,
        field: PaywallStateKey.Field,
        sourceComponent: PaywallComponentIdentity,
        sheetComponentID: String?
    ) {
        let key = PaywallStateKey.paywall(scope: self.scope, field: field)

>         self.store.request(

Author: High: The coordinator updates PackageContext by synchronously checking whether the store contains the originally requested package ID. That misses async gate resolution and replace(with:) cases. The spec requires gates to be able to replace mutations and resolve asynchronously, so the coordinator should observe committed root/sheet selection changes from the store and update PackageContext from committed values.

#endif
```

- [ ] **Step 6: Add override property key builder**

Create `RevenueCatUI/Templates/V2/State/PaywallOverridePropertyKeyBuilder.swift`:

```swift
import Foundation

#if !os(tvOS)

@available(iOS 15.0, macOS 12.0, tvOS 15.0, watchOS 8.0, *)
struct PaywallOverridePropertyKeyBuilder {

    func field(forPropertyPath propertyPath: String) -> PaywallStateKey.Field {
        .component(Self.normalize(propertyPath))
    }

    static func normalize(_ propertyPath: String) -> String {
        propertyPath
            .replacingOccurrences(of: "iconName", with: "icon_name")
            .replacingOccurrences(of: "backgroundColor", with: "background_color")
            .replacingOccurrences(of: "fontWeight", with: "font_weight")
            .replacingOccurrences(of: "fontName", with: "font_name")

>     }

Author: This is not needed. Whatever the value is after serialization is fine. If we convert to camel case during serialization, use camel case. If we have snake case, use snake case.

}

#endif
```

This starts with explicit normalization for existing camelCase coding keys. During implementation, prefer deriving names from each component's `CodingKeys` raw values so JSON names like `icon_name` and `background_color` become the state field names without extra SDK-authored mapping.

- [ ] **Step 7: Add reducer registry skeleton**

Create `RevenueCatUI/Templates/V2/State/PaywallStateReducerRegistry.swift`:

```swift
import Foundation

#if !os(tvOS)

@available(iOS 15.0, macOS 12.0, tvOS 15.0, watchOS 8.0, *)
struct PaywallStateReducerRegistry {

    private let keyBuilder = PaywallOverridePropertyKeyBuilder()

    func fields(forComponentType componentType: String, propertyNames: Set<String>) -> Set<PaywallStateKey.Field> {
        var fields = Set(propertyNames.map { self.keyBuilder.field(forPropertyPath: $0) })

        if componentType == "icon",
           propertyNames.contains("icon_name"),
           propertyNames.contains("formats") {
            fields.insert(.component("icon_url"))

>         }

Author: Here, can we use the coding key values instead of hard coded strings?
